### PR TITLE
fix: Generate useful error message when no expression on hook

### LINKED ISSUE #10910: An empty/missing expression in a non exit LifecycleHook produces a cryptic error message
### Pre-requisites

- [X] I have double-checked my configuration
- [X] I can confirm the issues exists when I tested with `:latest`
- [X] I'd like to contribute the fix myself (see [contributing guide](https://github.com/argoproj/argo-workflows/blob/master/docs/CONTRIBUTING.md))

### What happened/what you expected to happen?

When looking at #8954 the workflow example given wouldn't work because it has a missing `expression` in the non-exit hook.

The error you get is `unable to evaluate expression '': unexpected token EOF (1:1)` which is technically correct but completely unhelpful as to why you're wrong and what to do about it.

### Version

latest

### Paste a small workflow that reproduces the issue. We must be able to run the workflow; don't enter a workflows that uses private images.

```YAML
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: foo-
spec:
  entrypoint: main
  templates:
    - name: main
      steps:
        - - name: step-1
            template: one
            hooks:
              foo:
                # expression: "true"
                template: two
    - name: one
      script:
        image: python:alpine3.6
        command: [ python ]
        source: |
          print("one")
    - name: two
      script:
        image: python:alpine3.6
        command: [ python ]
        source: |
          print("two")
```

### Logs from the workflow controller

```text
time="2023-04-14T10:52:40.011Z" level=info msg="Created pod: foo-6hxnx[0].step-1 (foo-6hxnx-one-1793798762)" namespace=argo workflow=foo-6hxnx
time="2023-04-14T10:52:40.011Z" level=info msg="Created pod: foo-6hxnx[0].step-1 (foo-6hxnx-one-1793798762)" namespace=argo workflow=foo-6hxnx
time="2023-04-14T10:52:40.011Z" level=error msg="Mark error node" error="unable to evaluate expression '': unexpected token EOF (1:1)" namespace=argo nodeName="foo-6hxnx[0]" workflow=foo-6hxnx
time="2023-04-14T10:52:40.011Z" level=error msg="Mark error node" error="unable to evaluate expression '': unexpected token EOF (1:1)" namespace=argo nodeName="foo-6hxnx[0]" workflow=foo-6hxnx
time="2023-04-14T10:52:40.011Z" level=info msg="node foo-6hxnx-688708416 phase Running -> Error" namespace=argo workflow=foo-6hxnx
time="2023-04-14T10:52:40.011Z" level=info msg="node foo-6hxnx-688708416 message: unable to evaluate expression '': unexpected token EOF (1:1)" namespace=argo workflow=foo-6hxnx
time="2023-04-14T10:52:40.011Z" level=info msg="node foo-6hxnx-688708416 finished: 2023-04-14 10:52:40.011461263 +0000 UTC" namespace=argo workflow=foo-6hxnx
time="2023-04-14T10:52:40.011Z" level=info msg="node foo-6hxnx-688708416 phase Running -> Error" namespace=argo workflow=foo-6hxnx
time="2023-04-14T10:52:40.011Z" level=info msg="node foo-6hxnx-688708416 message: unable to evaluate expression '': unexpected token EOF (1:1)" namespace=argo workflow=foo-6hxnx
time="2023-04-14T10:52:40.011Z" level=info msg="node foo-6hxnx-688708416 finished: 2023-04-14 10:52:40.011461263 +0000 UTC" namespace=argo workflow=foo-6hxnx
time="2023-04-14T10:52:40.011Z" level=info msg="Workflow step group node foo-6hxnx-688708416 not yet completed" namespace=argo workflow=foo-6hxnx
time="2023-04-14T10:52:40.011Z" level=info msg="Workflow step group node foo-6hxnx-688708416 not yet completed" namespace=argo workflow=foo-6hxnx
```

### Logs from in your workflow's wait container

```text
Not relevant
```

### PR DESCRIPTION
**Pull request by Joibel**
The error message for a missing or empty expression was `unable to evaluate expression '': unexpected token EOF (1:1)`. This changes it to a more useful `Expression required for hook <hook path>`.

Also attempt to validate to prevent this problem upfront.

Fixes #10910

### CODE REVIEW

**Review by terrytangyuan — COMMENTED** [review]
Is this ready for review?

**Comment by Joibel** [comment]
Not yes, the validator isn't quite done.

File: workflow/controller/hooks.go [hunk 1]
```diff
@@ -18,12 +19,15 @@ func (woc *wfOperationCtx) executeWfLifeCycleHook(ctx context.Context, tmplCtx *
 		if hookName == wfv1.ExitLifecycleEvent {
 			continue
 		}
+		hookNodeName := generateLifeHookNodeName(woc.wf.ObjectMeta.Name, string(hookName))
```
Comment (terrytangyuan) [inline workflow/controller/hooks.go:22]: This is not needed until inside the execute block?

Comment (Joibel) [reply workflow/controller/hooks.go:22]: I use it on new line 24, two lines below this one as part of the error message.

**Review by terrytangyuan — APPROVED** [review]

**Pull request merged**